bottlecap/src/main.rs

```diff
@@ -156,23 +165,76 @@ fn main() -> Result<()> {
-                return Ok(());
+                shutdown = true;
             }
             Err(err) => {
                 eprintln!("Error: {:?}", err);
                 println!("Exiting");
                 return Err(err);
             }
         }
+        // Block until we get something from the telemetry API
+        // Check if flush logic says we should block and flush or not
+        if flush_control.should_flush() {
+            loop {
+                let received = event_bus.rx.recv();
+                if let Ok(event) = received {
+                    match event {
+                        Event::Metric(event) => {
+                            debug!("Metric event: {:?}", event);
+                        }
+                        Event::Telemetry(event) => {
+                            match event.record {
+                                TelemetryRecord::PlatformInitReport {
+                                    initialization_type,
+                                    phase,
+                                    metrics,
+                                } => {
+                                    debug!("Platform init report for initialization_type: {:?} with phase: {:?} and metrics: {:?}", initialization_type, phase, metrics);
+                                    // write this straight to metrics aggr
+                                    // write this straight to logs aggr
+                                    // write this straight to trace aggr
+                                }
+                                TelemetryRecord::PlatformRuntimeDone {
+                                    request_id, status, ..
+                                } => {
+                                    debug!(
+                                        "Runtime done for request_id: {:?} with status: {:?}",
+                                        request_id, status
+                                    );
+                                    debug!("FLUSHING ALL");
+                                    dogstats_client.flush();
+                                    debug!("CALLING FOR NEXT EVENT");
+                                    break;
+                                }
+                                TelemetryRecord::PlatformReport {
+                                    request_id, status, ..
+                                } => {
+                                    debug!(
+                                        "Platform report for request_id: {:?} with status: {:?}",
+                                        request_id, status
+                                    );
+                                }
+                                _ => {}
+                            }
+                        }
+                        _ => {}
+                    }
+                } else {
+                    error!("could not get the event");
+                }
+            }
+        }
+        if shutdown {
+            dogstats_client.shutdown();
```

Comment:
why can't we do that when receiving the shutdown event in the main loop?

Reply:
`shutdown` is a long-poll response from the `/next` endpoint, which means until that point we are frozen.

I moved `shutdown` out of main loop and into this conditional after we block on the telemetry API so that we can receive the final logs and metrics (eg `report` line) from the telemetry API before we shutdown. I noticed this because on shutdown events, we never flushed the last metric/log.